#166058055 Implement token-based authentication and authorization

elikeyz opened the pull request:
#### What does this PR do?
Setup token based authorization for all the routes except `signup` and `signin` and create user-specific routes
#### Description of Task to be completed?
- Setup token based authorization for all routes except `signup` and `signin`
- Create a route to get the logged in user's details
- Create a route to get the logged in user's loans
- Create a route to get the logged in user's repayments
#### How should this be manually tested?
Clone this repo, install the dependencies with the command `npm install`, start the app with `npm start` and navigate to `localhost:3000` on Postman
#### What are the relevant pivotal tracker stories?
[#166058055](https://www.pivotaltracker.com/n/projects/2326596)

elikeyz committed:
refactor(modal): refactor the display modal function

- refactor the display modal function by splitting it into two functions

[Finishes #166058055]

elikeyz committed:
feature(auth): implement token-based authorization

- use json web tokens to protect all the routes except the signup and
  signin routes

[Finishes #166058055]

elikeyz committed:
feature(user): create an endpoint to get user details

- create an endpoint to get the details of the logged in user

[Finishes #166058055]

elikeyz committed:
feature(user): create an endpoint to get user's loans

- create an endpoint to get all the loans requested by the logged in user

[Finishes #166058055]

elikeyz committed:
feature(user): create an endpoint to get user's loan repayments

- create an endpoint to get all the loan repayments of the logged in user

[Finishes #166058055]

Danielshow commented on src/test/repaymentsTest.js:
```diff
@@ -2,15 +2,48 @@
 
 describe('Repayments', () => {
   describe('GET /loans/:loanId/repayments', () => {
+    it('it should fail if there is no token in the header', (done) => {
```
Your test description does not necessarily need `it`. Your it blocks is doing that anyways
```suggestion
    it('should fail if there is no token in the header', (done) => {
```

elikeyz replied:
I've made the changes

Danielshow commented on src/utils/validators/tokenValidate.js:
```diff
@@ -0,0 +1,22 @@
+dotenv.config();
+
+const tokenValidate = (req, res, next) => {
+  if (!req.headers.token || !req.headers.token.trim()) {
```
The Token cannot be in the headers, it will be in the Authorization object in the header.
```suggestion
  if (!req.headers['authorization']) {
  else {
  const token = req.headers.authorization.split(' ')[1]
```

Danielshow replied:
You can check this article [link](https://medium.com/@maison.moa/using-jwt-json-web-tokens-to-authorize-users-and-protect-api-routes-3e04a1453c3e)

elikeyz replied:
I've made the changes

Danielshow reviewed (commented):
Attend to these comments

elikeyz committed:
chore(password): encrypt the passwords in the user data model

- encrypt the passwords for the user data objects stored in the data
  structure in the model
- use bcrypt to compare the encrypted passwords with the inputted password
  in the login middleware

[Finishes #166058055]

elikeyz committed:
chore(auth): require the auth token from the authorization object

- require the auth token from the authorization object in the header
- edit the test case description strings

[Finishes #166058055]
